Back-patch libpq support for TLS versions beyond v1

danielgustafsson opened the pull request:
This backports the frontend parts of TLSv1+ support from upstream, we already have the backend portions from an unrelated backport into libpq.
```
  commit 4dddf8552801ef013c40b22915928559a6fb22a0
  Author: Tom Lane <tgl@sss.pgh.pa.us>
  Date:   Thu May 21 20:41:55 2015 -0400

    Back-patch libpq support for TLS versions beyond v1.

    Since 7.3.2, libpq has been coded in such a way that the only SSL protocol
    it would allow was TLS v1.  That approach is looking increasingly obsolete.
    In commit 820f08cabdcbb899 we fixed it to allow TLS >= v1, but did not
    back-patch the change at the time, partly out of caution and partly because
    the question was confused by a contemporary server-side change to reject
    the now-obsolete SSL protocol v3.  9.4 has now been out long enough that
    it seems safe to assume the change is OK; hence, back-patch into 9.0-9.3.

    (I also chose to back-patch some relevant comments added by commit
    326e1d73c476a0b5, but did *not* change the server behavior; hence, pre-9.4
    servers will continue to allow SSL v3, even though no remotely modern
    client will request it.)

    Per gripe from Jan Bilek.
```

Eulerizeit commented:
@lij55 can you have someone look at this?

jchampio commented on src/backend/libpq/be-secure.c:
```diff
@@ -768,6 +768,13 @@ initialize_SSL(void)
+		 * We use SSLv23_method() because it can negotiate use of the highest
+		 * mutually supported protocol version, while alternatives like
+		 * TLSv1_2_method() permit only one specific version.  Note that we
+		 * don't actually allow SSL v2, only v3 and TLS protocols (see below).
```
Do you know the reason that the backend allows SSL3? Seems... dangerous.

jchampio replied:
Oh, I missed the last paragraph of the backported commit message. We may want to make sure that supporting SSLv3 doesn't open us up to downgrade attacks; I can't remember if those were fixed or not.

lij55 commented:
Clients team will review this today.

danielgustafsson commented:
Backported another small commit to remove SSLv3 support from the backend while we're at it.

danielgustafsson commented:
This failed in the PR pipeline with:
```======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_Remote_basic|Test Details|Basic RemoteOperation test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 21, in test_Remote_basic
    self.assertTrue(TestOperation().run() == RemoteOperation(TestOperation(), "localhost").run())
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_Remote_harden|Test Details|Ensure that some logging occurs in event of error.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 70, in test_Remote_harden
    RemoteOperation(RaiseOperation_Unpicklable(), "localhost").run()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_inner_exceptions|Test Details|Verify that an object not at the global level of this file cannot be pickled properly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 31, in test_inner_exceptions
    RemoteOperation(RaiseOperation_Nested(), "localhost").run()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_proper_exceptions_sanity|Test Details|
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 49, in test_proper_exceptions_sanity
    RemoteOperation(RaiseOperation_Safe(), "localhost").run()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_proper_exceptions_with_args|Test Details|
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 57, in test_proper_exceptions_with_args
    RemoteOperation(RaiseOperation_Safe(), "localhost").run()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

======================================================================
ERROR: Test Suite Name|operations.test.unit.test_unit_utils|Test Case Name|test_unsafe_exceptions_with_args|Test Details|
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py", line 40, in test_unsafe_exceptions_with_args
    RemoteOperation(RaiseOperation_Unsafe(), "localhost").run()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/tmp/build/edb35eda/gpdb_src/gpMgmt/bin/gppylib/operations/utils.py", line 43, in execute
    execname = os.path.split(sys.argv[0])[-1]
IndexError: list index out of range

----------------------------------------------------------------------
```
I would happily appreciate any help in deciphering what that actually means..

pivotal-mike commented:
@danielgustafsson Can we rebase re-run this though the PR pipeline?  This looks lie something fixed last week

danielgustafsson commented:
Pushed to master after a green PR run, thanks
